recipes/lora_finetune_distributed_td.py

```diff
@@ -0,0 +1,1119 @@
+
+        cfg_datasets = cfg_dataloader["datasets"]
+        weights, datasets = {}, {}
+        if isinstance(cfg_datasets, DictConfig):
+            for k, cfg_and_weight in cfg_datasets.items():
+                weights[k] = float(cfg_and_weight.get("weight", 1.0))
+                datasets[k] = Prefetcher(
+                    self._setup_one_dataset(
+                        cfg_dataset=cfg_and_weight.dataset,
+                        global_shuffle=shuffle,
+                        global_parallel_method=parallel_method,
+                        global_streaming=streaming,
+                        global_num_workers=num_workers,
+                    ),
+                    prefetch_factor=8,
+                )
+
+        # Instantiate collate_fn
+        if "left_pad_sequence" in collate_fn:
+            raise RuntimeError("left_pad_sequence collator is only for inference.")
+        collate_fn = _get_component_from_path(collate_fn)
+
+        # TODO: add multi-dataset mixer
+        if num_workers == 0:
+            _Mapper = Mapper
+        else:
+            _Mapper = partial(
+                ParallelMapper,
+                num_workers=num_workers,
+                method=parallel_method,
+            )
+        if len(cfg_datasets) == 1:
+            node = next(iter(datasets.values()))  # TODO: multi dataset
+        else:
+            node = MultiDatasetWeightedSampler(
+                source_nodes=datasets,
+                weights=weights,
+            )
+        node = Batcher(node, batch_size, drop_last=True)
+        node = _Mapper(
+            node,
+            map_fn=(
+                partial(
+                    collate_fn,  # noqa
+                    padding_idx=self._tokenizer.pad_id,
+                    ignore_idx=self._loss_fn.ignore_index,
+                )
+                if not packed
+                else padded_collate_packed
+            ),
+        )
+        if pin_memory:
+            node = PinMemory(node)
+        if num_workers > 0:
+            node = Prefetcher(node, 2)
+
+        log.info("TorchData nodes are initialized")
+
+        return node
+
+    def _setup_data(
+        self,
+        cfg_dataset: DictConfig,
+        shuffle: bool,
+        batch_size: int,
+        collate_fn: str,
+    ) -> Tuple[DistributedSampler, DataLoader]:
+        """
+        All data related setup happens here. Currently this recipe only supports the
+        DistributedSamplers with Map-style Datasets which fit into memory. Other samplers,
+        iterable datasets and streaming datasets are not supported.
+        """
+        world_size, rank = training.get_world_size_and_rank()
+
+        if isinstance(cfg_dataset, ListConfig):
+            datasets = [
+                config.instantiate(single_cfg_dataset, self._tokenizer)
+                for single_cfg_dataset in cfg_dataset
+            ]
+            ds = ConcatDataset(datasets=datasets)
+            packed = False
```

Comment:
We can add a log here or raise an error in the init in case packed was set to true. i.e. we dont want to silently override the config